File: 0200_numIslands.py
from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        number_of_islands = 0
        visited = set()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    if (i, j) not in visited:
                        queue = deque([(i, j)])
                        visited.add((i, j))
                        while queue:
                            r, c = queue.popleft()
                            for row, col in [r - 1, c], [r + 1, c], [r, c - 1], [r, c + 1]:
                                if 0 <= row < m and 0 <= col < n and grid[row][col] == '1' and (row, col) not in visited:
                                    queue.append((row, col))
                                    visited.add((row, col))
                        number_of_islands += 1
        return number_of_islands
        
# Alternative solution

class Solution1:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        number_of_islands = 0
        visited = set()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    if (i, j) not in visited:
                        self.dfs(grid, visited, i, j, m, n)
                        number_of_islands += 1
        return number_of_islands
    
    def dfs(self, grid, visited, i, j, m, n):
        if (i, j) not in visited:
            visited.add((i, j))
            for row, col in [i - 1, j], [i + 1, j], [i, j - 1], [i, j + 1]:
                if 0 <= row < m and 0 <= col < n and grid[row][col] == '1' and (row, col) not in visited:
                    self.dfs(grid, visited, row, col, m, n)

# Alternative solution

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        numOfIslands = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    numOfIslands += 1
                    self.traverseIsland(grid, m, n, i, j)
        return numOfIslands

    def traverseIsland(self, matrix, m, n, row, col):
        if m > row >= 0 <= col < n and matrix[row][col] == '1':
            matrix[row][col] = '0'
            self.traverseIsland(matrix, m, n, row, col + 1)
            self.traverseIsland(matrix, m, n, row + 1, col)
            self.traverseIsland(matrix, m, n, row, col - 1)
            self.traverseIsland(matrix, m, n, row - 1, col)

File: 0501-1000/0200_numIslands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int numOfIslands = 0;
        Queue<int[]> queue;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    queue = new LinkedList<>();
                    queue.add(new int[] {i, j});
                    grid[i][j] = '@';
                    numOfIslands++;

                    while (!queue.isEmpty()) {
                        int[] coordinate = queue.remove();
                        int x = coordinate[0], y = coordinate[1];

                        for (int[] direction: directions) {
                            int r = direction[0], c = direction[1];

                            if (0 <= x+r && x+r < m && 0 <= y+c && y+c < n && grid[x+r][y+c] == '1') {
                                grid[x+r][y+c] = '@';
                                queue.add(new int[] {x+r, y+c});
                            }
                        }
                    }
                }
            }
        }

        return numOfIslands;
    }
}
